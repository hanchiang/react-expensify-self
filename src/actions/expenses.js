import {
  ADD_EXPENSE, START_ADD_EXPENSE, 
  EDIT_EXPENSE, START_EDIT_EXPENSE,
  REMOVE_EXPENSE, START_REMOVE_EXPENSE,
  SET_EXPENSE, START_SET_EXPENSE
} from '../constants/actionTypes';

const addExpense = (expense) => ({
  type: ADD_EXPENSE,
  expense
});

const startAddExpense = (expense) => ({
  type: START_ADD_EXPENSE,
  expense
});

const editExpense = (id, updates) => ({
  type: EDIT_EXPENSE,
  id,
  updates
});

const startEditExpense = (id, updates) => ({
  type: START_EDIT_EXPENSE,
  id,
  updates
});

const removeExpense = (id) => ({
  type: REMOVE_EXPENSE,
  id
});

const startRemoveExpense = (id) => ({
  type: START_REMOVE_EXPENSE,
  id
});

const setExpense = (expenses) => ({
  type: SET_EXPENSE,
  expenses
});

const startSetExpense = () => ({
  type: START_SET_EXPENSE
});

export { 
  addExpense, startAddExpense, editExpense, startEditExpense,
  removeExpense, startRemoveExpense, setExpense, startSetExpense
};