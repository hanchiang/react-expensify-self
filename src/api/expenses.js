import { database } from '../firebase/firebase';

const addToDatabase = (path, expense) => {
  return database.ref(path).push(expense).then(ref => ({
    id: ref.key,
    ...expense
  }));
};

const fetchFromDatabase = (path) => {
  let expenses = [];

  return database.ref(path).once('value').then(snapshot => {
    let expenses = [];
    snapshot.forEach(childSnapshot => {
      expenses.push({
        id: childSnapshot.key,
        ...childSnapshot.val()
      });
    });
    return expenses;
  });
}

const editToDataBase = (path, updates) => database.ref(path).update(updates);
const removeFromDatabase = (path) => database.ref(path).remove();

export { addToDatabase, fetchFromDatabase, editToDataBase, removeFromDatabase };