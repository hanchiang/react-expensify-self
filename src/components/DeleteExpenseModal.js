import React from 'react';
import Modal from 'react-modal';

Modal.setAppElement('#app');

function DeleteExpenseModal(props) {
  const handleConfirmDelete = (event) => props.onModalConfirm(true);
  const handleRejectDelete = (event) => props.onModalConfirm(false);

  return (
    <Modal
      isOpen={props.showModal}
      onRequestClose={props.onModalClose}
      closeTimeoutMS={300}
      contentLabel="Delete expense?"
      className="modal"
    >
      <div className="modal-header">
        <h2 className="modal-title">Delete Expense</h2>
      </div>

      <div className="modal-body">
        <p className="modal-message">Are you sure you want to delete expense?</p>
      </div>

      <div className="modal-actions">
        <button className="button button-normal" onClick={handleRejectDelete} type="button">Close</button>
        <button className="button button-warning" onClick={handleConfirmDelete} type="button">Delete</button>  
      </div>
    </Modal>
  );

}

export default DeleteExpenseModal;