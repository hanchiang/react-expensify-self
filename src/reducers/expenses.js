import { ADD_EXPENSE, EDIT_EXPENSE, REMOVE_EXPENSE, SET_EXPENSE } from '../constants/actionTypes';

/*
expense = {
  id: 1,
  description: 'hello',
  note: '', (optional)
  createdAt: 123
}
*/

const initialState = [];

function expensesReducer(state = initialState, action) {
  switch(action.type) {
    case ADD_EXPENSE:
      return [...state, action.expense];
    case EDIT_EXPENSE:
      return state.map(expense => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates
          };
        } else {
          return expense;
        }
      });
    case REMOVE_EXPENSE:
      return state.filter(expense => expense.id !== action.id);
    case SET_EXPENSE:
      return action.expenses
    default:
      return state;
  }
}

export default expensesReducer;