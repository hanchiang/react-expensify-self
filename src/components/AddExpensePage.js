import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';
import moment from 'moment';

export function AddExpensePage(props) {
  const onSubmit = (expense) => {
    props.onSubmit(expense);
    props.history.push('/dashboard');
  }

  return (
    <div>
      <div className="page-header">
        <div className="content-container">
          <h1 className="page-header-title">Add Expense</h1>
        </div>
      </div>
      
      <div className="content-container">
        <ExpenseForm onSubmit={onSubmit} />
      </div>
      
    </div>
  );
}

const mapDispatchToProps = (dispatch) => ({
  onSubmit: (expense) => dispatch(startAddExpense(expense))
});

export default connect(null, mapDispatchToProps)(AddExpensePage);