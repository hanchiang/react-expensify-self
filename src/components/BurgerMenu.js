import React from 'react';
import { slide as Menu } from 'react-burger-menu';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogout } from '../actions/auth';

export class BurgerMenu extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isOpen: false
    };
  }

  isMenuOpen(state) {
    return state.isOpen;
  }

  render() {
    return (
      <Menu right isOpen={this.state.isOpen} onStateChange={this.isMenuOpen}>
          <Link to="/profile">Profile</Link>
          <button className="button button-link logout-button" type="button" onClick={this.props.startLogout}>Logout</button>
      </Menu>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(null, mapDispatchToProps)(BurgerMenu);