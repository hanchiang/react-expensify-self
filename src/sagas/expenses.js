import { database } from '../firebase/firebase';
import { select, put, call } from 'redux-saga/effects';
import React from 'react';
import ReactDOM from 'react-dom';

import { renderApp } from '../app';
import { addExpense, setExpense, editExpense, removeExpense } from '../actions/expenses';
import { addToDatabase, fetchFromDatabase, editToDataBase, removeFromDatabase } from '../api/expenses';


function* handleAddExpense(action) {
  const { expense } = action;
  const state = yield select();
  const uid = state.auth.uid;

  const path = `users/${uid}/expenses`;
  const addedExpense = yield call(addToDatabase, path, expense);
  yield put(addExpense(addedExpense));
}

function* handleSetExpense(action) {
  const state = yield select();
  const uid = state.auth.uid;
  const path = `users/${uid}/expenses`;
  const expenses = yield call(fetchFromDatabase, path);
  yield put(setExpense(expenses));
  yield renderApp();
}

function* handleEditExpense(action) {
  const { updates, id } = action;
  const state = yield select();
  const uid = state.auth.uid;
  const path = `users/${uid}/expenses/${id}`;
  yield call(editToDataBase, path, updates);
  yield put(editExpense(id, updates));
}

function* handleRemoveExpense(action) {
  const { id } = action;
  const state = yield select();
  const uid = state.auth.uid;
  const path = `users/${uid}/expenses/${id}`;
  yield call(removeFromDatabase, path);
  yield put(removeExpense(id));
}

export { handleAddExpense, handleSetExpense, handleEditExpense, handleRemoveExpense };