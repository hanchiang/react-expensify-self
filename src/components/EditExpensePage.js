import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import DeleteExpenseModal from './DeleteExpenseModal';

class EditExpensePage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      showModal: false,
      shouldRemoveExpense: false
    };
    this.onSubmit = this.onSubmit.bind(this);
    this.onRemoveClick = this.onRemoveClick.bind(this);
    this.onModalClose = this.onModalClose.bind(this);
    this.onModalConfirm = this.onModalConfirm.bind(this);
  }
  onSubmit(expense) {
    console.log('submit!')
    this.props.onSubmit(expense);
    this.props.history.push('/dashboard');
  }

  onRemoveClick() {
    this.setState({ showModal: true });
  }

  onModalClose() {
    this.setState({ showModal: false });
  }

  onModalConfirm(shouldRemoveExpense) {
    this.setState({ shouldRemoveExpense, showModal: false });
    if (shouldRemoveExpense) {
      this.props.onRemove(this.props.expense.id);
      this.props.history.push('/dashboard');
    }
  }

  render() {
    const { expense } = this.props;
    return (
      <div id="edit-expense-page">
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header-title">Edit Expense</h1>
          </div>
        </div>

        <div className="content-container">
          <ExpenseForm expense={expense} onSubmit={this.onSubmit} isEdit={true} onRemove={this.onRemoveClick} />
        </div>

      {
        <DeleteExpenseModal showModal={this.state.showModal}
          onModalConfirm={this.onModalConfirm}
          onModalClose={this.onModalClose}
        />
      }
      </div>
    );
  }
  
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  onSubmit: (expense) => dispatch(startEditExpense(props.match.params.id, expense)),
  onRemove: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);