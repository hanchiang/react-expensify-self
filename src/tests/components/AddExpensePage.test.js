import React from 'react';
import { shallow } from 'enzyme';

import { AddExpensePage } from '../../components/AddExpensePage';
import expenses from '../fixtures/expenses';

describe('AddExpensePage', () => {
  it('should render AddExpensePage', () => {
    const wrapper = shallow(<AddExpensePage />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should trigger onSubmit correctly', () => {
    const onSubmit = jest.fn();
    const history = {push: jest.fn()};
    const wrapper = shallow(<AddExpensePage onSubmit={onSubmit} history={history} />);
    // console.log(wrapper.find('ExpenseForm').prop('onSubmit'))
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
    expect(onSubmit).toHaveBeenCalledWith(expenses[0]);
    expect(history.push).toHaveBeenCalledWith('/dashboard');
  });
});
