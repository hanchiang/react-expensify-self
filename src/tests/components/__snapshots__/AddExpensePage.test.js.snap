// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddExpensePage should render AddExpensePage 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header-title"
      >
        Add Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
